### PR TITLE
Implement support for RSVP and sales support

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: https://central.tri.be/issues/64669

This PR updates the RSVP tickets code to:

* not increase sales when an RSVP is Not Going
* increase sales when Going
* allow a robust filtering of the RSVP states with variable stock sizes
* tests for all the code (integration as it's as small as I could go)
* add some filters to prevent email and anywhere I felt we were lacking

### CODE REVIEW

File: src/Tribe/RSVP.php [hunk 1]
```diff
@@ -340,7 +346,8 @@ public function update_attendee_data( $data, $order_id, $event_id ) {
 		}
 
 		// Dont try to Save if it's restricted
-		if ( ! isset( $attendee['product_id'] ) || Tribe__Tickets__Tickets_View::instance()->is_rsvp_restricted( $event_id, $attendee['product_id'] ) ) {
+		if ( ! isset( $attendee['product_id'] ) || $this->tickets_view->is_rsvp_restricted( $event_id,
+		                                                                               $attendee['product_id'] ) ) {
```
Comment (borkweb) [inline src/Tribe/RSVP.php]: Formatting of this conditional is funky. Maybe:

```php
if (
  ! isset( $attendee['product_id'] )
  || $this->tickets_view->is_rsvp_restricted( $event_id, $attendee['product_id'] )
) {
```

Comment (lucatume) [inline src/Tribe/RSVP.php]: I blame PHPStorm.

File: src/Tribe/RSVP.php [hunk 2]
```diff
@@ -534,31 +563,58 @@ public function generate_tickets( ) {
...
-		 * @param int    $event_id ID of the event the order was placed for
+		 * @param int    $order_id              ID of the RSVP order
+		 * @param int    $post_id               ID of the post the order was placed for
 		 * @param string $attendee_order_status 'yes' if the user indicated they will attend
 		 */
-		do_action( 'event_tickets_rsvp_tickets_generated', $order_id, $event_id, $attendee_order_status );
+		do_action( 'event_tickets_rsvp_tickets_generated', $order_id, $post_id, $attendee_order_status );
+
+		// @todo: this positive stati should be filterable
+		$send_mail_stati = array('yes');
```
Comment (borkweb) [inline src/Tribe/RSVP.php]: `array( 'yes' )`

Also: It _looks_ like you have a filter for this down below. Is this comment still needed?

**Comment by tr1b0t** [comment]
Status|Message|Line
:-----:|-----|-----
:warning: | Expected next thing to be an escaping function (see Codex for 'Data Validation'), not '$html' | [Attendance_Totals.php:61](https://github.com/moderntribe/event-tickets/blob/ad4832808b1045d35d07bd0dfad17dfe4348abfc/src/Tribe/RSVP/Attendance_Totals.php#L61)

> via jenkins/codesniffer-phpcs

File: tests/integration/Tribe/Tickets/Tickets_ViewTest.php [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+			'yes-plus-one'    => 'Going +1',
+			'yes-plus-family' => 'Going +family',
+			'yes-with-mt'     => 'Going (with MT)',
+		];
+
+		add_filter(
+			'event_tickets_rsvp_options', function ( $options ) use ( $rsvp_options ) {
+
+			return array_merge( $options, $rsvp_options );
+		}
```
Comment (borkweb) [inline tests/integration/Tribe/Tickets/Tickets_ViewTest.php]: Spacing is all funky with this `add_filter()`

File: tests/integration/Tribe/Tickets/Tickets_ViewTest.php [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
```
Comment (borkweb) [inline tests/integration/Tribe/Tickets/Tickets_ViewTest.php]: Spacing funkiness

Comment (borkweb) [inline tests/integration/Tribe/Tickets/Tickets_ViewTest.php]: Rinse and repeat on all the `add_filters()` in this file

File: tests/integration/Tribe/Tickets/Tickets_ViewTest.php [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+	 * it should prune RSVP options that do not have right format
+	 */
+	public function it_should_prune_rsvp_options_that_do_not_have_right_format() {
+		$rsvp_options = [
+			// good
+			'yes-plus-one'    => [ 'label' => 'Going +1', 'decrease_stock_by' => 2 ],
+			// good even without stock
+			'maybe'           => [ 'label' => 'Maybe' ],
+			// no label
+			'yes-plus-family' => [ 'Going +family' ],
```
Comment (borkweb) [inline tests/integration/Tribe/Tickets/Tickets_ViewTest.php]: Isn't this an allowed type as per [this comment](https://github.com/moderntribe/event-tickets/pull/259/files#diff-d9386274c70c90592678be7875280994R523)?

Comment (lucatume) [inline tests/integration/Tribe/Tickets/Tickets_ViewTest.php]: No, if it is an array it has to define the `label` else just supply the label

Comment (lucatume) [inline tests/integration/Tribe/Tickets/Tickets_ViewTest.php]: And the comment is wrong...

**Comment by borkweb** [comment]
This looks good once you finish fixing the `add_filter()` spacing funkiness in `Tickets_ViewTest.php`. I'll preemptively add a *merge* label here.

Ⓜ️

File: src/Tribe/RSVP/Attendance_Totals.php [hunk 6]
```diff
@@ -42,13 +42,23 @@ public function print_totals() {
 		$going = $this->get_total_going();
 		$not_going = $this->get_total_not_going();
 
-		echo "
+		$html = "
 			<ul>
 				<li> <strong>$total_rsvps_label</strong> $total_rsvps </li>
 				<li> <strong>$going_label</strong> $going </li>
 				<li> <strong>$not_going_label</strong> $not_going </li>
+				$attendance_lines_after
```
Comment (barryhughes) [inline src/Tribe/RSVP/Attendance_Totals.php]: Hey @lucatume - just came across this while investigating a different problem. Right now, this variable is undefined: was it intended to be a spot for devs to inject their own totals, or something like that?

Comment (lucatume) [inline src/Tribe/RSVP/Attendance_Totals.php]: It was meant exactly for that but leaving it undefined was not my intent and is a left-over mistake by my side.

Comment (lucatume) [inline src/Tribe/RSVP/Attendance_Totals.php]: Nope, ignore my previous comment: the filter is in place (the `tribe_tickets_rsvp_print_totals_html` one) and that's a left-over that var is a leftover that should be removed.

**Pull request merged**